### PR TITLE
Add retries option for consuming from an exclusively-consumed queue

### PR DESCRIPTION
**Pull request by baelter**
## Background

A queue with an exclusive consumer refuses a second one with `ACCESS_REFUSED` (403). During a rolling deploy the new process typically starts consuming before the old one has released the queue, so callers had to wrap `subscribe` in their own retry loop to survive the handover. Nothing is registered for auto-recovery until the first consume succeeds, so a refused initial attempt could not be picked up by the reconnect supervisor either. The Ruby client solves the same problem in cloudamqp/amqp-client.rb#135; this is the equivalent for the JS client, plus recovery coverage.

## Summary

`AMQPQueue.subscribe()` takes two new params: `retries` (default `0`, unchanged behavior) and `retryDelay` (default 1000 ms). A consume refused with 403 is retried that many times at that interval before the original error is rethrown.

```javascript
const sub = await q.subscribe({ exclusive: true, retries: 5 }, async (msg) => { /* ... */ })
```

The retries are stored on the consumer definition, so consumer recovery after a reconnect retries too. Without that, a reconnect landing while another process still holds the exclusive consumer reports the failure through `onrecoverfailed` and leaves the subscription dead until the next disconnect.

`AMQPError` now carries `code`, the AMQP reply code from a broker channel or connection close (403, 404, ...). The retry check compares that instead of matching the error message text, and callers can branch on it too.

## Tests

Four tests against a live broker, using a second connection to hold the exclusive consumer: default subscribe fails immediately, a retrying subscribe succeeds once the holder lets go, retries exhausted rejects with the original error, and recovery after a reconnect retries until the holder releases (asserting the retry path actually ran, not a lucky race).

### CODE REVIEW

File: src/amqp-base-client.ts [hunk 1]
```diff
@@ -424,7 +424,7 @@ export abstract class AMQPBaseClient {
                   this.logger?.debug("channel", channelId, "closed", code, text, classId, methodId)
 
                   const msg = `channel ${channelId} closed: ${text} (${code})`
-                  const err = new AMQPError(msg, this)
+                  const err = new AMQPError(msg, this, code)
                   channel.setClosed(err)
```
[the remark this answers is not in the document]
Comment (baelter) [reply src/amqp-base-client.ts:428]: Fixed in 0a5f647. `AMQPChannel.onerror` now takes the reply code as a second argument, `setClosed` passes `err.code`, and the default handler forwards it into the error it hands to `connection.onerror`. Covered by a new test in test/test.ts that asserts 403 on both the rejected promise and the connection handler.

File: src/amqp-queue.ts [hunk 2]
```diff
@@ -378,7 +396,30 @@ export class AMQPQueue<
...
   }
 
+  /**
+   * Open a consumer, retrying while another connection holds the queue's
+   * exclusive consumer. Nothing is registered for recovery until the first
+   * consume succeeds, so a refused initial attempt can only be retried here.
+   */
   private async openConsumer(def: ConsumerDefinition): Promise<AMQPConsumer | AMQPGeneratorConsumer> {
+    const retries = def.retries ?? 0
+    const retryDelay = def.retryDelay ?? 1000
```
[the remark this answers is not in the document]
Comment (baelter) [reply src/amqp-queue.ts:407]: Fixed in 0a5f647. `subscribe` validates up front: `retries` must be a non-negative safe integer (so NaN and Infinity throw), `retryDelay` a finite non-negative number. New test covers all four cases.

File: src/amqp-queue.ts [hunk 3]
```diff
@@ -378,7 +396,30 @@ export class AMQPQueue<
...
+      try {
+        return await this.consume(def)
+      } catch (err) {
+        if (attempt >= retries || !isAccessRefused(err)) throw err
+        this.session.logger?.debug(
+          `Queue ${this.name} has an exclusive consumer, retrying in ${retryDelay}ms ` +
+            `(attempt ${attempt + 1}/${retries})`,
+        )
+        await new Promise((resolve) => setTimeout(resolve, retryDelay))
+        if (this.session.stopping || this.session.closed) throw err
```
[the remark this answers is not in the document]
Comment (baelter) [reply src/amqp-queue.ts]: Fixed in 0a5f647. The wait goes through a new internal `AMQPSession.sleep()` that registers the waiter, and `stop()` wakes every waiter and clears its timer, so a pending retry no longer holds the process open.

File: test/amqp-session.ts [hunk 4]
```diff
@@ -1712,3 +1712,106 @@ test("onblocked / onunblocked can be wired through options", async () => {
...
+test("consumer recovery retries while another connection holds the exclusive consumer", async () => {
+  const name = "test-exclusive-recover-" + Math.random()
+  const recoverFailed = vi.fn()
+  const debug = vi.fn()
+  const received = new Promise<string>((resolve) => {
+    void withSession(
+      async (session) => {
+        const q = await session.queue(name, { durable: false, autoDelete: false })
+        const body = new Promise<string>((gotMessage) => {
+          void q
```
[the remark this answers is not in the document]
Comment (baelter) [reply test/amqp-session.ts]: Fixed in 0a5f647. The test is linear now: connect, subscribe, destroy the socket, grab the holder, publish, then await the delivery. Nothing is floated except the timed `holder.close()`, and setup failures reject the test instead of hanging.

File: test/amqp-session.ts [hunk 5]
```diff
@@ -1712,3 +1712,106 @@ test("onblocked / onunblocked can be wired through options", async () => {
...
+
+test("subscribe rejects when the consume retries are exhausted", async () => {
+  const name = "test-exclusive-exhausted-" + Math.random()
+  const holder = await holdExclusiveConsumer(name)
+  try {
+    await withSession(async (session) => {
+      const q = await session.queue(name, { durable: false, autoDelete: false })
+      await expect(q.subscribe({ exclusive: true, retries: 2, retryDelay: 50 }, () => {})).rejects.toThrow(
+        /ACCESS_REFUSED/,
+      )
```
[the remark this answers is not in the document]
Comment (baelter) [reply test/amqp-session.ts]: Fixed in 0a5f647. The test asserts the exact retry log attempts, `["1/2", "2/2"]`, so exactly two retries have to happen before the rejection.

File: test/amqp-session.ts [hunk 6]
```diff
@@ -1712,3 +1712,132 @@ test("onblocked / onunblocked can be wired through options", async () => {
...
+    const received = new Promise<void>((resolve) => (gotMessage = resolve))
+    await q.subscribe({ exclusive: true, retries: 20, retryDelay: 100 }, (msg) => {
+      bodies.push(msg.bodyString() ?? "")
+      gotMessage?.()
+    })
+
+    // Drop the connection, then grab the freed exclusive consumer from another
+    // connection so recovery hits ACCESS_REFUSED and has to retry.
+    ;(testClient(session) as AMQPClient).socket?.destroy()
+    holder = await holdExclusiveConsumer(name)
```
[the remark this answers is not in the document]
Comment (baelter) [reply test/amqp-session.ts]: Fixed in 3e21c79. The holder is now installed from an awaited `beforeConnect` hook on the reconnect attempt, so it is always in place before the session reconnects, and `holdExclusiveConsumer` retries until the broker has released the previous holder. The holder is released only after a retry has actually been logged (`vi.waitFor`), so no step depends on timing. Ran it 5x in a row plus the full suite, green.

**Comment by baelter** [comment]
Both suppressed comments from the last review are fixed in 4b72096: `AMQPSession.sleep()` returns immediately when the session is already stopped, and the timed `holder.close()` in the exclusive-consumer test swallows its rejection.

**Comment by ThomasSarlin** [comment]
Questions:
- Edge case, 0 retryDelay coupled with a large number for retries makes for a quite neat loop to hammer the broker. Does it make sense with some minimal value or just leave as is?

- The retry triggers on any 403, but ACCESS_REFUSED also covers permission errors like if the user lacks permission to read a queue, not just the exclusive-consumer case. So a subscribe that can never succeed (bad permissions) will burn through all retries before failing. Is that acceptable?
